classes/aioseop_logging.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+/**
+ * @package All-in-One-SEO-Pack
+ */
+
+/**
+ * Class aioseop_logging
+ *
+ * Class used for logging.
+ */
+// @codingStandardsIgnoreStart
+class aioseop_logging {
+// @codingStandardsIgnoreEnd
+
+	const LOG_LENGTH = 100;
+	const LOG_EXPIRY_MINS = 5;
+	static $__log = array();
+	static $log_types = array( 'error', 'warn', 'info', 'debug' );
+
+	private $log_length = self::LOG_LENGTH;
+	private $log_expiry = self::LOG_EXPIRY_MINS;
+
+	function __construct() {
+		$this->log_length = apply_filters( 'aioseop_log_length', self::LOG_LENGTH );
+		$this->log_expiry = apply_filters( 'aioseop_log_expiry_minutes', self::LOG_EXPIRY_MINS );
+
+		// use the action to extend behavior.
+		add_action( 'aiosp_log', array( $this, 'log' ), 10, 5 );
+		// update db only once.
+		add_action( 'shutdown', array( $this, 'update' ) );
+	}
+
+	function update() {
+		error_log("shutdown");
+		$__log = self::$__log;
+		if ( count( $__log ) > $this->log_length ) {
+			$__log = array_slice( $__log, 0 - $this->log_length );
+		}
+		error_log("log " . print_r($__log,true));
+
+		set_transient( 'aiosp_event_log', $__log, $this->log_expiry * MINUTE_IN_SECONDS );
+	}
+
+	// log to the database and into file.
+	function log( $log_msg, $log_type = 'debug', $file = null, $line = null, $force = false ) {
+		global $aioseop_options;
+
+		// first check if logging is enabled?
+		if ( ! ( ! empty( $aioseop_options ) && isset( $aioseop_options['aiosp_do_log'] ) && $aioseop_options['aiosp_do_log'] ) ) {
+			if ( ! ( $force || defined( 'AIOSEOP_DO_LOG' ) ) ) {
+				return;
+			}
+		}
+
+		if ( ! in_array( $log_type, self::$log_types, true ) ) {
+			return;
+		}
+
+		$time = date( 'F j, Y H:i:s', current_time( 'timestamp', true ) );
+
+		if ( is_array( $log_msg ) || is_object( $log_msg ) ) {
+			$log_msg = print_r( $log_msg, true );
+		}
+
+		self::$__log[] = array(
+			'type' => $log_type,
+			'msg'  => $log_msg,
+			'time' => $time,
+			'file' => $file,
+			'line' => $line,
+		);
+
+		$log_line = sprintf( '%s(%s) - %s:%d - %s', $time, $log_type, $file, $line, $log_msg );
```

Comment:
sprintf can be taxing, and there's no translations going on here. You could just concat a string.

Reply:
From what I have researched, I am not sure about this claim. check one of the many: https://stackoverflow.com/questions/7147305/performance-of-variable-expansion-vs-sprintf-in-php

Reply:
They also say the same thing. I don't see why sprintf is needed when there is no translation. Was just worried considering how dynamic and large the log string could be.